docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,471 @@
+  call `OptimizerDef.apply_param_gradient()` (implemented by subclasses of
+  `OptimizerDef`).
+- The functions `init_param_state()` and `apply_param_gradient()` are called
+  for every leaf in the params/grads pytree. This makes it possible to write the
+  calculations directly without `jax.tree_multimap()`.
+- The interface was defined in pre-Linen without the distinction of `params` vs.
+  other collections in `variables` in mind. The original API was elegant because
+  one only needed to pass around the optimizer, which included the parameters,
+  optimizer state, optimizer hyperparameters, and a reference to the
+  `OptimizerDef` to perform the param/state update.
+
+## Previous Training Step
+[Previous Training Step]: #previous-training-step
+
+An optimizer would first be constructed from its definition and the pytree of
+target params:
+
+```python
+optimizer_def = flax.optim.Momentum(learning_rate=0.1, beta=0.9)
+optimizer = optimizer_def.create(variables['params'])
+```
+
+Then, the target variables would optimized in the train step (assuming a single
+non-params collection "batch_stats"):
+
+```python
+def make_train_step(apply_fn):
+  @jax.jit
+  def train_step(optimizer, batch_stats, inputs, labels):
+
+    def loss_fn(params):
+      variables = {'params': params, 'batch_stats': batch_stats}
+      logits, new_model_state = apply_fn(
+          variables, inputs, mutable=['batch_stats'])
+      loss = xent_loss(logits, labels)
+      return loss, new_model_state['batch_stats']
+
+    (loss, new_batch_stats), grad = jax.value_and_grad(loss_fn, has_aux=True)(
+        optimizer.target)
+    lr = get_learning_rate(step)
+    new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+    return new_optimizer, new_batch_stats, loss
+
+  return train_step
+
+
+batch_stats = variables['batch_stats']
+train_step = make_train_step(model.apply)
+for step, batch in enumerate(ds)
+  optimizer, batch_stats, loss = train_step(
+      optimizer, batch_stats, batch['image'], batch['label'])
+```
+
+Remarks:
+
+- Notice how `optimizer.apply_gradients()` can take additional arguments to
+  update hyperparameters, such as learning rate from an independent function
+  `get_learning_rate()` in this case.
+
+
+# Update Plan
+[Update Plan]: #update-plan
+
+1. Finalize discussions on this FLIP
+2. Test existing optimizers for numerical equivalence (e.g. `flax.optim.Adam`
+   and `optax.adamw`).
+3. Update examples to use Optax and verify that they reach the same final
+   performance with the same computational cost. We probably want some examples
+   to directly use the optimizer while others might use [Linen helper].
+4. Port missing optimizers to Optax (e.g. Adafactor) - and verify above points.
+5. Update all documentation (including README, Flax guided tour, HOWTOs, ...) to
+   talk exclusively about Optax optimizers.
+6. Mark optimizers in `flax.optim` as deprecated.
+
```

Comment:
I think we also need:
* A transition guide, similar to the [one for Linen](https://docs.google.com/document/d/1hYavTVPaKVVe9Be8pCB7yW7r6dDv3RALVNit8NZca4c/edit), with common patterns in the old and new form
* As part of this transition guide and (2), document how to translate hyperparameters from current flax.optim to Optax (including default values) -- @mtthss is working on this here: https://github.com/deepmind/optax/blob/e186087ddce5ef92d5df936ff341e819b29bc576/optax/_src/equivalence_test.py#L100

Reply:
Good idea - added.